src/bin/cobalt/serve.rs

```diff
@@ -78,6 +77,69 @@ pub fn serve_command(matches: &clap::ArgMatches) -> Result<()> {
     Ok(())
 }
 
+fn static_file_handler(dest: &path::Path, req: Request) -> Result<()> {
+    // grab the requested path
+    let mut req_path = req.url().to_string();
+
+    // strip off any querystrings so path.is_file() matches and doesn't stick index.html on the end
+    // of the path (querystrings often used for cachebusting)
+    if let Some(position) = req_path.rfind('?') {
+        req_path.truncate(position);
+    }
+
+    // find the path of the file in the local system
+    // (this gets rid of the '/' in `p`, so the `join()` will not replace the path)
+    let path = dest.to_path_buf().join(&req_path[1..]);
+
+    let serve_path = if path.is_file() {
+        // try to point the serve path to `path` if it corresponds to a file
+        path
+    } else {
+        // try to point the serve path into a "index.html" file in the requested
+        // path
+        path.join("index.html")
+    };
+
+    // if the request points to a file and it exists, read and serve it
+    if serve_path.exists() {
+        let mut file = fs::File::open(serve_path)?;
+        // buffer to store the file
+        let mut buffer: Vec<u8> = vec![];
+        file.read_to_end(&mut buffer)?;
+        req.respond(Response::from_data(buffer))?;
+    } else {
+        // write a simple body for the 404 page
+        req.respond(
+            Response::from_string("<h1> <center> 404: Page not found </center> </h1>")
+                .with_status_code(404),
+        )?;
+    }
+    Ok(())
+}
+
+fn serve(dest: &path::Path, ip: &str) -> Result<()> {
+    info!("Serving {:?} through static file server", dest);
+    info!("Server Listening on http://{}", &ip);
+    info!("Ctrl-c to stop the server");
+
+    // attempts to create a server
+    let server = match Server::http(ip) {
+        Ok(server) => server,
+        Err(e) => {
+            error!("{}", e);
+            process::exit(1);
+        }
+    };
+
+    for request in server.incoming_requests() {
+        if let Err(e) = static_file_handler(&dest, request) {
+            error!("{}", e);
+            process::exit(1);
+        }
+    }
```

Comment:
ditto about propagating up

Reply:
Oh, you are right that this should ignore errors because we want to keep running in case updating the site failed once